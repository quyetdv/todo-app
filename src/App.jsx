import { useState, useMemo, useEffect } from 'react';
import { Button } from 'antd';
import './App.css';
import TodoTable from './components/TodoTable';
import TodoListContext from './context/TodoList';
import TodoItemModal from './components/TodoItemModal';

const defaultTask = {
  description: '',
  task: '',
  priority: 0,
  key: 0,
  completed: false,
};

function App() {
  const [todoList, setTodoList] = useState(
    JSON.parse(localStorage.getItem('todoList')) || []
  );

  // only rerun context when todoList is updated instead of on every render
  const value = useMemo(
    () => ({
      todoList,
      setTodoList,
    }),
    [todoList]
  );

  useEffect(() => {
    // storing data after changes registered
    localStorage.setItem('todoList', JSON.stringify(todoList));
  }, [todoList]);

  const [visible, setVisible] = useState(false);
  const [isEdit, setIsEdit] = useState(false);
  const [currentTask, setCurrentTask] = useState(defaultTask);

  const onConfirm = (values, key) => {
    if (isEdit) {
      // update the list
      const updatedList = todoList.map((todoItem) => {
        if (todoItem.key === key) {
          return { ...todoItem, ...values };
        }
        return todoItem;
      });
      setTodoList(updatedList);
    } else {
      // add new task to existing list
      const existingData = JSON.parse(JSON.stringify(todoList));
      setTodoList([
        ...existingData,
        { ...values, key: +new Date(), completed: false },
      ]);
    }
    setVisible(false);
  };

  const onView = (key) => {
    const selectedTask = todoList.find((todoItem) => todoItem.key === key);
    setCurrentTask(selectedTask);
    setIsEdit(true);
    setVisible(true);
  };

  const onMarkDone = (key) => {
    const updatedList = todoList.map((todoItem) => {
      if (todoItem.key === key) {
        return { ...todoItem, completed: true };
      }
      return todoItem;
    });
    setTodoList(updatedList);
  };

  const onDelete = (key) => {
    const taskToDelete = todoList.findIndex((todoItem) => todoItem.key === key);
    const clonedList = JSON.parse(JSON.stringify(todoList));
    clonedList.splice(taskToDelete, 1);
    setTodoList(clonedList);
  };

  return (
    <TodoListContext.Provider value={value}>
      <div className="App">
        <header className="App-header">
          <Button
            type="primary"
            onClick={() => {
              setCurrentTask(defaultTask);
              setIsEdit(false);
              setVisible(true);
            }}
          >
            New Task
          </Button>

          <TodoTable
            onView={onView}
            onMarkDone={onMarkDone}
            onDelete={onDelete}
          />

          <TodoItemModal
            visible={visible}
            onConfirm={onConfirm}
            onCancel={() => {
              setVisible(false);
            }}
            isEdit={isEdit}
            currentTask={currentTask}
          />
        </header>
      </div>
    </TodoListContext.Provider>
  );
}

export default App;
