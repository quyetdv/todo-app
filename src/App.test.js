import { render, fireEvent, waitFor } from '@testing-library/react';
import React from 'react';
import renderer from 'react-test-renderer';
import App from './App';

describe('App', () => {
  beforeAll(() => {
    Object.defineProperty(window, 'matchMedia', {
      writable: true,
      value: jest.fn().mockImplementation((query) => ({
        matches: false,
        media: query,
        onchange: null,
        addListener: jest.fn(), // deprecated
        removeListener: jest.fn(), // deprecated
        addEventListener: jest.fn(),
        removeEventListener: jest.fn(),
        dispatchEvent: jest.fn(),
      })),
    });
  });

  it('renders App component', () => {
    const component = renderer.create(<App />);

    expect(component.toJSON()).toMatchSnapshot();
  });

  it('creates a new task', async () => {
    const { queryByText, getByLabelText } = render(<App />);

    fireEvent.click(queryByText('New Task'));
    await waitFor(async () => {
      expect(queryByText('Creating a new task')).toBeInTheDocument();
      const taskInput = getByLabelText('Task');
      fireEvent.change(taskInput, { target: { value: 'Task number 1' } });
      fireEvent.click(queryByText('Confirm'));
      await waitFor(() => {
        expect(queryByText('Task number 1')).toBeInTheDocument();
      });
    });

    fireEvent.click(queryByText('New Task'));
    await waitFor(async () => {
      expect(queryByText('Creating a new task')).toBeInTheDocument();
      const taskInput = getByLabelText('Task');
      fireEvent.change(taskInput, { target: { value: 'Task number 2' } });
      fireEvent.click(queryByText('Confirm'));
      await waitFor(() => {
        expect(queryByText('Task number 2')).toBeInTheDocument();
      });
    });

    fireEvent.click(queryByText('New Task'));
    await waitFor(async () => {
      expect(queryByText('Creating a new task')).toBeInTheDocument();
      const taskInput = getByLabelText('Task');
      fireEvent.change(taskInput, { target: { value: 'Task number 4' } });
      fireEvent.click(queryByText('Confirm'));
      await waitFor(() => {
        expect(queryByText('Task number 4')).toBeInTheDocument();
      });
    });
  });

  it('edit existing task', async () => {
    const { getByText, getAllByText, getByLabelText } = render(<App />);

    fireEvent.click(getAllByText('Edit')[0]);
    await waitFor(async () => {
      expect(getByText('Editing current task')).toBeInTheDocument();
      const taskInput = getByLabelText('Task');
      fireEvent.change(taskInput, { target: { value: 'Task number 3' } });
      fireEvent.click(getByText('Confirm'));
      await waitFor(() => {
        expect(getByText('Task number 3')).toBeInTheDocument();
      });
    });
  });

  it('stop using modal by clicking Cancel', async () => {
    const { getByText, getAllByText } = render(<App />);

    fireEvent.click(getAllByText('Edit')[0]);
    await waitFor(async () => {
      expect(getByText('Editing current task')).toBeInTheDocument();
      fireEvent.click(getByText('Cancel'));
    });
  });

  it('marks a task as done', async () => {
    const { getAllByText } = render(<App />);

    const firstDoneButton = getAllByText('Mark As Done')[0];
    fireEvent.click(firstDoneButton);
    expect(firstDoneButton.closest('button')).toBeDisabled();
  });

  it('deletes a task', async () => {
    const { getAllByText, queryByText } = render(<App />);

    const firstDoneButton = getAllByText('Delete')[0];
    fireEvent.click(firstDoneButton);
    expect(queryByText('Task number 3')).not.toBeInTheDocument();
  });
});
