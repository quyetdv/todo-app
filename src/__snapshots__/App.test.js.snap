// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`App renders App component 1`] = `
<div
  className="App"
>
  <header
    className="App-header"
  >
    <button
      className="ant-btn ant-btn-primary"
      onClick={[Function]}
      type="button"
    >
      <span>
        New Task
      </span>
    </button>
    <div
      className="ant-table-wrapper todo-table"
    >
      <div
        className="ant-spin-nested-loading"
      >
        <div
          className="ant-spin-container"
        >
          <div
            className="ant-table ant-table-bordered ant-table-empty"
          >
            <div
              className="ant-table-container"
            >
              <div
                className="ant-table-content"
                onScroll={[Function]}
                style={Object {}}
              >
                <table
                  style={
                    Object {
                      "tableLayout": "auto",
                    }
                  }
                >
                  <colgroup>
                    <col
                      style={
                        Object {
                          "width": "30%",
                        }
                      }
                    />
                    <col
                      style={
                        Object {
                          "width": "10%",
                        }
                      }
                    />
                    <col
                      style={
                        Object {
                          "width": "40%",
                        }
                      }
                    />
                    <col
                      style={
                        Object {
                          "width": "20%",
                        }
                      }
                    />
                  </colgroup>
                  <thead
                    className="ant-table-thead"
                  >
                    <tr>
                      <th
                        className="ant-table-cell ant-table-column-has-sorters"
                        colSpan={null}
                        onClick={[Function]}
                        onMouseEnter={[Function]}
                        onMouseLeave={[Function]}
                        rowSpan={null}
                        style={Object {}}
                      >
                        <div
                          className="ant-table-column-sorters"
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                        >
                          <span
                            className="ant-table-column-title"
                          >
                            Task
                          </span>
                          <span
                            className="ant-table-column-sorter ant-table-column-sorter-full"
                          >
                            <span
                              className="ant-table-column-sorter-inner"
                            >
                              <span
                                aria-label="caret-up"
                                className="anticon anticon-caret-up ant-table-column-sorter-up"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="caret-up"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="0 0 1024 1024"
                                  width="1em"
                                >
                                  <path
                                    d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                  />
                                </svg>
                              </span>
                              <span
                                aria-label="caret-down"
                                className="anticon anticon-caret-down ant-table-column-sorter-down"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="caret-down"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="0 0 1024 1024"
                                  width="1em"
                                >
                                  <path
                                    d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </span>
                        </div>
                      </th>
                      <th
                        className="ant-table-cell ant-table-column-has-sorters"
                        colSpan={null}
                        onClick={[Function]}
                        onMouseEnter={[Function]}
                        onMouseLeave={[Function]}
                        rowSpan={null}
                        style={Object {}}
                      >
                        <div
                          className="ant-table-column-sorters"
                          onMouseEnter={[Function]}
                          onMouseLeave={[Function]}
                        >
                          <span
                            className="ant-table-column-title"
                          >
                            Priority
                          </span>
                          <span
                            className="ant-table-column-sorter ant-table-column-sorter-full"
                          >
                            <span
                              className="ant-table-column-sorter-inner"
                            >
                              <span
                                aria-label="caret-up"
                                className="anticon anticon-caret-up ant-table-column-sorter-up"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="caret-up"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="0 0 1024 1024"
                                  width="1em"
                                >
                                  <path
                                    d="M858.9 689L530.5 308.2c-9.4-10.9-27.5-10.9-37 0L165.1 689c-12.2 14.2-1.2 35 18.5 35h656.8c19.7 0 30.7-20.8 18.5-35z"
                                  />
                                </svg>
                              </span>
                              <span
                                aria-label="caret-down"
                                className="anticon anticon-caret-down ant-table-column-sorter-down"
                                role="img"
                              >
                                <svg
                                  aria-hidden="true"
                                  data-icon="caret-down"
                                  fill="currentColor"
                                  focusable="false"
                                  height="1em"
                                  viewBox="0 0 1024 1024"
                                  width="1em"
                                >
                                  <path
                                    d="M840.4 300H183.6c-19.7 0-30.7 20.8-18.5 35l328.4 380.8c9.4 10.9 27.5 10.9 37 0L858.9 335c12.2-14.2 1.2-35-18.5-35z"
                                  />
                                </svg>
                              </span>
                            </span>
                          </span>
                        </div>
                      </th>
                      <th
                        className="ant-table-cell"
                        colSpan={null}
                        onMouseEnter={[Function]}
                        onMouseLeave={[Function]}
                        rowSpan={null}
                        style={Object {}}
                      >
                        Description
                      </th>
                      <th
                        className="ant-table-cell"
                        colSpan={null}
                        onMouseEnter={[Function]}
                        onMouseLeave={[Function]}
                        rowSpan={null}
                        style={Object {}}
                      >
                        Action
                      </th>
                    </tr>
                  </thead>
                  <tbody
                    className="ant-table-tbody"
                  >
                    <tr
                      className="ant-table-placeholder"
                      style={
                        Object {
                          "display": null,
                        }
                      }
                    >
                      <td
                        className="ant-table-cell"
                        colSpan={4}
                        onMouseEnter={[Function]}
                        onMouseLeave={[Function]}
                        rowSpan={null}
                        style={Object {}}
                      >
                        <div
                          className="ant-empty ant-empty-normal"
                        >
                          <div
                            className="ant-empty-image"
                          >
                            <svg
                              className="ant-empty-img-simple"
                              height="41"
                              viewBox="0 0 64 41"
                              width="64"
                              xmlns="http://www.w3.org/2000/svg"
                            >
                              <g
                                fill="none"
                                fillRule="evenodd"
                                transform="translate(0 1)"
                              >
                                <ellipse
                                  className="ant-empty-img-simple-ellipse"
                                  cx="32"
                                  cy="33"
                                  rx="32"
                                  ry="7"
                                />
                                <g
                                  className="ant-empty-img-simple-g"
                                  fillRule="nonzero"
                                >
                                  <path
                                    d="M55 12.76L44.854 1.258C44.367.474 43.656 0 42.907 0H21.093c-.749 0-1.46.474-1.947 1.257L9 12.761V22h46v-9.24z"
                                  />
                                  <path
                                    className="ant-empty-img-simple-path"
                                    d="M41.613 15.931c0-1.605.994-2.93 2.227-2.931H55v18.137C55 33.26 53.68 35 52.05 35h-40.1C10.32 35 9 33.259 9 31.137V13h11.16c1.233 0 2.227 1.323 2.227 2.928v.022c0 1.605 1.005 2.901 2.237 2.901h14.752c1.232 0 2.237-1.308 2.237-2.913v-.007z"
                                  />
                                </g>
                              </g>
                            </svg>
                          </div>
                          <div
                            className="ant-empty-description"
                          >
                            No Data
                          </div>
                        </div>
                      </td>
                    </tr>
                  </tbody>
                </table>
              </div>
            </div>
            <div
              className="ant-table-footer"
            >
              <span
                className="footer-span"
              >
                Total: 
                0
              </span>
              <span
                className="footer-span"
              >
                Completed: 
                0
              </span>
            </div>
          </div>
        </div>
      </div>
    </div>
  </header>
</div>
`;
